nemo_rl/environments/search/search.py

```diff
@@ -0,0 +1,340 @@
+                    final_result = "\n---\n".join(pretty_results)
+                    result_text = json.dumps({"result": final_result})
+                else:
+                    result_text = json.dumps({"result": "No search results found."})
+            except Exception as e:
+                error_msg = f"Error processing search results: {e}"
+                result_text = json.dumps({"result": error_msg})
+        else:
+            result_text = json.dumps(
+                {"result": "Unknown API state (no response and no error message)."}
+            )
+        return result_text
+
+    def _is_done(self, log: LLMMessageLogType, metadata: SearchEnvMetadata):
+        if metadata["num_turns"] >= self.max_turns:
+            return True
+        msg = log[-1]["content"]
+        return "<answer>" in msg and "</answer>" in msg
+
+    def _get_reward(self, answer: str | None, ground_truth: str, done: bool):
+        if done:
+            return compute_score(answer, ground_truth)
+        else:
+            return 0.0
+
+    def _parse_action(self, action: str) -> List[Optional[str]]:
+        match = None
+        if "<search>" in action and "</search>" in action:
+            match = re.search(r"<search>(.*?)</search>", action, re.DOTALL)
+        return match.group(1) if match else None
+
+    def process_turn(self, log: LLMMessageLogType, metadata: SearchEnvMetadata):
+        ground_truth = metadata["ground_truth"]
+
+        done = self._is_done(log, metadata)
+        answer = extract_solution(log[-1]["content"])
+        reward = self._get_reward(answer, ground_truth, done)
+        error = None
+        if done:
+            return SearchWorkerResult(
+                observation={
+                    "role": "user",
+                    "content": "Correct" if reward > 0.0 else "Incorrect",
+                },
+                reward=reward,
+                done=done,
+                error=None,
+                metadata=metadata,
+                answer=answer,
+            )
+        try:
+            query = self._parse_action(log[-1]["content"])
+            if query:
+                obs = self.search(query)
+            else:
+                obs = None
+                error = "No search query provided."
+        except Exception as e:
+            error = str(e)
+            obs = None
+
+        if obs:
+            new_msg = {
+                "role": "user",
+                "content": "<information>" + obs + "</information>",
+            }
+        else:
+            new_msg = {"role": "user", "content": "<error>" + error + "</error>"}
+
+        new_metadata = metadata.copy()
+        new_metadata["num_turns"] += 1
+
+        return SearchWorkerResult(
+            observation=new_msg,
+            reward=reward,
+            done=done,
+            error=error,
+            metadata=new_metadata,
+            answer=answer,
+        )
+
+
+@ray.remote
+class SearchEnv(EnvironmentInterface[SearchEnvMetadata]):
```

Comment:
We should probably add unit test for this. I'm suspecting that it's not gonna be easy to actually bring up a 2-GPU retrival server for unit tests, hence you might need to mock `requests.session` for unit testing this.